sycl/include/sycl/reduction.hpp

```diff
@@ -8,16 +8,2464 @@
 
 #pragma once
 
-#if __cplusplus >= 201703L
-// Entire feature is dependent on C++17.
-
+#include <sycl/accessor.hpp>
+#include <sycl/atomic.hpp>
+#include <sycl/atomic_ref.hpp>
+#include <sycl/detail/tuple.hpp>
+#include <sycl/ext/oneapi/accessor_property_list.hpp>
+#include <sycl/group_algorithm.hpp>
+#include <sycl/handler.hpp>
+#include <sycl/kernel.hpp>
 #include <sycl/known_identity.hpp>
+#include <sycl/properties/reduction_properties.hpp>
+
+#include <tuple>
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace detail {
 
-#include "sycl/ext/oneapi/reduction.hpp"
+/// Base non-template class which is a base class for all reduction
+/// implementation classes. It is needed to detect the reduction classes.
+class reduction_impl_base {};
 
+/// Predicate returning true if all template type parameters except the last one
+/// are reductions.
+template <typename FirstT, typename... RestT> struct AreAllButLastReductions {
+  static constexpr bool value =
+      std::is_base_of<reduction_impl_base,
+                      std::remove_reference_t<FirstT>>::value &&
+      AreAllButLastReductions<RestT...>::value;
+};
+
+/// Helper specialization of AreAllButLastReductions for one element only.
+/// Returns true if the template parameter is not a reduction.
+template <typename T> struct AreAllButLastReductions<T> {
+  static constexpr bool value =
+      !std::is_base_of<reduction_impl_base, std::remove_reference_t<T>>::value;
+};
+} // namespace detail
+} // __SYCL_INLINE_VER_NAMESPACE(_V1)
+} // namespace sycl
+
+#if __cplusplus >= 201703L
+// Entire feature is dependent on C++17. We still have to make the trait above
+// available as queue shortcuts use them unconditionally, including on
+// non-reduction path.
 namespace sycl {
 __SYCL_INLINE_VER_NAMESPACE(_V1) {
 
+
+/// Class that is used to represent objects that are passed to user's lambda
+/// functions and representing users' reduction variable.
+/// The generic version of the class represents those reductions of those
+/// types and operations for which the identity value is not known.
+/// The View template describes whether the reducer owns its data or not: if
+/// View is 'true', then the reducer does not own its data and instead provides
+/// a view of data allocated elsewhere (i.e. via a reference or pointer member);
+/// if View is 'false', then the reducer owns its data. With the current default
+/// reduction algorithm, the top-level reducers that are passed to the user's
+/// lambda contain a private copy of the reduction variable, whereas any reducer
+/// created by a subscript operator contains a reference to a reduction variable
+/// allocated elsewhere. The Subst parameter is an implementation detail and is
+/// used to spell out restrictions using 'enable_if'.
+template <typename T, class BinaryOperation, int Dims, size_t Extent,
+          bool View = false, typename Subst = void>
+class reducer;
```

Comment:
This was moved from `sycl::ext::oneapi::detail` to `sycl` because

1) That's what standard seems to say
2) To make it work with ESIMD processing. We need it disallowed and everything in `sycl::detail` is allowed, while only an explicit subset of `sycl` is allowed in there.